pkg/controller/istiocontrolplane/listeners.go

```diff
@@ -0,0 +1,264 @@
+package istiocontrolplane
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"istio.io/operator/pkg/apis/istio/v1alpha1"
+	"istio.io/operator/pkg/helmreconciler"
+)
+
+const (
+	// ChartOwnerKey is the annotation key used to store the name of the chart that created the resource
+	ChartOwnerKey = MetadataNamespace + "/chart-owner"
+)
+
+// IstioListenerFactory is the factory used to create the rendering listeners for IstioControlPlane resources
+type IstioListenerFactory struct{}
+
+var _ helmreconciler.RenderingListenerFactory = &IstioListenerFactory{}
+
+// IstioRenderingListener is a RenderingListener specific to IstioControlPlane resources
+type IstioRenderingListener struct {
+	*helmreconciler.CompositeRenderingListener
+}
+
+// NewRenderingListener returns a new IstioRenderingListener, which is a composite that includes IstioStatusUpdater
+// and IstioChartCustomizerListener.
+func (f *IstioListenerFactory) NewRenderingListener(instance runtime.Object) (helmreconciler.RenderingListener, error) {
+	if istioControlPlane, ok := instance.(*v1alpha1.IstioControlPlane); ok {
+		return &IstioRenderingListener{
+			CompositeRenderingListener: &helmreconciler.CompositeRenderingListener{
+				Listeners: []helmreconciler.RenderingListener{
+					NewChartCustomizerListener(),
+					NewIstioStatusUpdater(istioControlPlane),
+				},
+			},
+		}, nil
+	}
+	return nil, fmt.Errorf("object is not an IstioControlPlane")
+}
+
+// IstioStatusUpdater is a RenderingListener that updates the status field on the IstioControlPlane
+// instance based on the results of the Reconcile operation.
+type IstioStatusUpdater struct {
+	*helmreconciler.DefaultRenderingListener
+	instance   *v1alpha1.IstioControlPlane
+	reconciler *helmreconciler.HelmReconciler
+}
+
+var _ helmreconciler.RenderingListener = &IstioStatusUpdater{}
+var _ helmreconciler.ReconcilerListener = &IstioStatusUpdater{}
+
+// NewIstioStatusUpdater returns a new IstioStatusUpdater instance for the specified IstioControlPlane
+func NewIstioStatusUpdater(instance *v1alpha1.IstioControlPlane) helmreconciler.RenderingListener {
+	return &IstioStatusUpdater{
+		DefaultRenderingListener: &helmreconciler.DefaultRenderingListener{},
+		instance:                 instance,
+	}
+}
+
+// EndReconcile updates the status field on the IstioControlPlane instance based on the resulting err parameter.
+func (u *IstioStatusUpdater) EndReconcile(_ runtime.Object, err error) error {
+	status := u.instance.Status
+	if err == nil {
+		if condition := status.GetCondition(v1alpha1.ConditionTypeInstalled); condition.Status != v1alpha1.ConditionStatusTrue {
```

Comment:
This could be greatly simplified if we move to a simpler enum based state machine.

Reply:
fine with me.  i can look into that.